implementations/rust/ockam/ockam_vault/src/vault/vault_builder.rs

```diff
@@ -0,0 +1,119 @@
+#[cfg(feature = "storage")]
+use crate::storage::PersistentStorage;
+use crate::vault::secrets_store_impl::VaultSecretsStore;
+use crate::{
+    AsymmetricVault, Implementation, Kms, SecretsStore, Signer, SymmetricVault, Vault, VaultKms,
+    VaultStorage,
+};
+use ockam_core::compat::sync::Arc;
+#[cfg(feature = "storage")]
+use ockam_core::Result;
+use ockam_node::InMemoryKeyValueStorage;
+#[cfg(feature = "std")]
+use std::path::Path;
+
+/// Builder for Vaults
+/// The `VaultBuilder` allows the setting of different implementations for the external interfaces of a Vault:
+///   `SecretsStore`, `AsymmetricVault`, `SymmetricVault`, `Signer`.
+///
+/// It is important to note that the `AsymmetricVault`, `SymmetricVault` and `Signer` interfaces
+/// depend on a shared `SecretsStore` implementation for ephemeral and persistent secrets.
+/// So when setting specific implementations for these traits it is important that the implementations
+/// share consistent storages.
+pub struct VaultBuilder {
+    secrets_store: Arc<dyn SecretsStore>,
+    asymmetric_vault: Arc<dyn AsymmetricVault>,
+    symmetric_vault: Arc<dyn SymmetricVault>,
+    signer: Arc<dyn Signer>,
+}
+
+impl VaultBuilder {
+    pub(crate) fn new_builder() -> VaultBuilder {
+        let kms = VaultKms::create_with_storage(InMemoryKeyValueStorage::create());
+        let secrets_store = Arc::new(VaultSecretsStore::new(
+            kms.clone(),
+            InMemoryKeyValueStorage::create(),
+        ));
+        let asymmetric_vault = secrets_store.clone();
+        let symmetric_vault = secrets_store.clone();
+        let signer = secrets_store.clone();
+        Self {
+            secrets_store,
+            asymmetric_vault,
+            symmetric_vault,
+            signer,
+        }
+    }
+
+    /// Set a persistent storage as a file storage with a specific path
+    /// Note: this overrides all previously set implementations
+    #[cfg(feature = "std")]
+    pub async fn with_persistent_storage_path(&mut self, path: &Path) -> Result<&mut Self> {
```

Comment:
Right, I made top-level functions `Vault::create_with_xxx` to support the most frequent use cases so that the builder is only used if absolutely needed by other implementors to fine-tweak the setup.